README.md
# Blender Unreal Engine Workspace

![Blender Unreal Engine Workspace](https://anasrar.github.io/Blender-UE4-Workspace/img/blender-unreal-engine-4-workspace-banner.png)

Blender **2.8x** *(above)* add-on for export directly to Unreal Engine 4 with all setting in Blender (inspired by **send to unreal** add-on).

## Feature

Allow you export static mesh, skeletal mesh, and animation with single click directly to Unreal Engine 4 or to FBX file.

![Feature Node](https://anasrar.github.io/Blender-UE4-Workspace/img/feature-node.png "Feature Node")

and yeah, I don't really have any future plan. so if you have any suggest just open new issue.

## Main Feature

### Static Mesh

Export for Static Mesh.

- Export to FBX and Unreal Engine
- Custom collision from vertices
- Custom collision from mesh - ```v.1.2```
- Custom lightmaps
- Export profile - ```v.1.2```
- Socket System - ```v.1.3```
- Level of Detail - ```v.1.4```
- Import Static Mesh From Unreal Engine ```v.1.4```

### Character

Export for Character (Skeletal Mesh).

- Export to FBX and Unreal Engine  
- Modular character
- Skeleton preset (Epic skeleton)
- Add twist bone for skeleton preset - ```v.1.2```
- Generate rig for skeleton preset - ```v.1.2```
- Export profile - ```v.1.2```
- Socket System (**Not Support Export**) - ```v.1.3```
- Character Part Manager ```v.1.4```
- Import Skeletal Mesh From Unreal Engine ```v.1.4```

### Animation

Export for Animation.

- Export to FBX and Unreal Engine - ```v.1.2```  
- Export profile - ```v.1.2```
- Import Animation From Unreal Engine ```v.1.4```

### Retarget Animation (Experimental)

Retarget Animation to another skeleton - ```v.1.3``` **Experimental**, ```v.1.4``` **Production Ready**

## Documetation

[documentation page](https://anasrar.github.io/Blender-UE4-Workspace/) or [youtube playlist](https://www.youtube.com/playlist?list=PLolnhUV-ZzXrXx1gJunoknuni8klsy0wH)

![GitHub release (latest by date)](https://img.shields.io/github/v/release/anasrar/Blender-UE4-Workspace?style=flat-square)

## How it works

Unreal Engine 4 allow to remote execute python script, with that we can execute python script import assets (FBX File) to Unreal Engine 4.

![Blender Unreal Engine Workspace FlowChart](https://anasrar.github.io/Blender-UE4-Workspace/img/flowchart.png "Flowchart")

## Download

You can download from

- Gumroad for latest version
  - https://gumroad.com/l/BlenderUnrealEngineWorkspace

- Github for pervious version
  - https://github.com/anasrar/Blender-UE4-Workspace/releases

master branch is **unstable** and **bug fix version**

## Installation

You can watch this video https://www.youtube.com/watch?v=38d5Myrh3ic or simply follow this instruction below.

### Blender

Edit &nbsp; 🡆 &nbsp; Preferences &nbsp; 🡆 &nbsp; Add-ons &nbsp; 🡆 &nbsp; Install &nbsp; 🡆 &nbsp; Select **UE4Workspace.zip** &nbsp; 🡆 &nbsp; Install Add-ons

### Unreal Engine 4

Edit &nbsp; 🡆 &nbsp; Plugins &nbsp; 🡆 &nbsp; Type "Script" On Search Bar &nbsp; 🡆 &nbsp; Enabled **Python Editor Script Plugin** and **Editor Scripting Utilities** &nbsp; 🡆 &nbsp; Reset Project

Edit &nbsp; 🡆 &nbsp; Project Setting &nbsp; 🡆 &nbsp; Plugin &nbsp; 🡆 &nbsp; Python &nbsp; 🡆 &nbsp; Check **Enable Remote Execution**?

Then you can try to connect your project from blender

## Usage

Press **N** on Blender for open the tab menu.

## Compatibility Test

tested with blender version

* **2.81** - **DROP** from version `v.1.3`

* **2.82** - **DROP** from version `v.1.4`

* **2.83** (make sure export folder path is absolute)

* **2.90** (make sure export folder path is absolute)

using Blender latest version for better compatibility.

## Support

You can support me through Gumroad

any donation will be appreciated.

## Contributing

For major changes or features request, please open an issue first to discuss what you would like to change or add.

## Changelog

Any changelog in [documentation page](https://anasrar.github.io/Blender-UE4-Workspace/changelog/) 

## License

This project is licensed under the **GPL-3.0** License - see the [LICENSE](LICENSE) file for details

docs/changelog.md
---
title: Changelog
---

# Changelog

## v.1.4.0

- [x] FIX : wrong bone orient in blender 2.83.5
- [x] FIX : collision and socket not export when hide
- [x] FEATURE : snap IK and FK operator
- [x] FEATURE : retarget animation (ready for production)
- [x] FEATURE : import assets from unreal engine project
- [x] FEATURE : LOD system for static mesh
- [x] FEATURE : set pose to A-pose and apply pose on skeleton preset
- [x] FEATURE : character part (mesh) panel for manage export character part
- [X] MISC : change algorithm animation list

## v.1.3.0

- [x] FEATURE : socket
- [x] FEATURE : humanoid face rig
- [x] FEATURE : retarget animation (experimental)
- [X] MISC : improve control rig
- [X] MISC : improve custom collision
- [X] MISC : improve custom bone for skeleton preset
- [X] MISC : all object panel now using sub panel
- [X] MISC : panel closed by default

## v.1.2.1

- [x] FIX : wrong bone orient in blender 2.83
- [x] FEATURE : custom collision picker from mesh
- [x] FEATURE : export animation
- [x] FEATURE : export profile
- [x] FEATURE : generate rig
- [x] FEATURE : add twist bone
- [x] MISC : reset variable after open blend file (disconnect from unreal and character skeleton list)
- [x] MISC : optimize make custom collision
- [x] MISC : move custom collision into collections
- [x] MISC : make custom collision semi transparent green for better visibility

## v.1.0.1

- [x] Initial release

__init__.py
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTIBILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

bl_info = {
    "name" : "UE4Workspace",
    "author" : "Anas RAR",
    "description" : "Addon For UE4 Workspace",
    "blender" : (2, 81, 0),
    "version" : (1, 4, 0),
    "location" : "3D View > Tools",
    "warning" : "",
    "wiki_url": "https://github.com/anasrar/Blender-UE4-Workspace", # 2.82 below
    "doc_url": "https://github.com/anasrar/Blender-UE4-Workspace", # 2.83 above
    "tracker_url": "https://github.com/anasrar/Blender-UE4-Workspace/issues",
    "support": "COMMUNITY",
    "category" : "Workspace"
}

import bpy
from bpy.app.handlers import persistent
from bpy.props import (StringProperty, BoolProperty, BoolVectorProperty, IntProperty, IntVectorProperty, FloatProperty, EnumProperty, PointerProperty, CollectionProperty)
from bpy.utils import (register_class, unregister_class)

from . UE4WS_Preferences import (
    Groups as PreferencesGroup,
    Preferences,
    Ops as PreferencesOperator
)

from . UE4WS_ExportOption import (
    PANEL as exportOptionPanel,
    Ops as exportOptionOperator
)

from . UE4WS_ExportOptionImportAssets import (
    IMPORTASSETS_UL_AssetsList,
    Groups as exportOptionImportAssetsGroup,
    Props as exportOptionImportAssetsProps,
    PANEL as exportOptionImportAssetsPanel,
    Ops as exportOptionImportAssetsOperator
)

from . UE4WS_ExportOptionImportAssetsFBXOption import (
    PANEL as exportOptionImportAssetsFBXOptionPanel
)

from . UE4WS_ExportOptionImportAssetsUnrealEngine import (
    PANEL as exportOptionImportAssetsUnrealEnginePanel
)

from . UE4WS_Object import (
    PANEL as objectPanel,
    Ops as objectOperator
)

from . UE4WS_ObjectCustomCollision import (
    Props as objectCustomCollisionProps,
    PANEL as objectCustomCollisionPanel,
    Ops as objectCustomCollisionOperator
)

from . UE4WS_ObjectCharacterPart import (
    Props as objectCharacterPartProps,
    PANEL as objectCharacterPartPanel,
    Ops as objectCharacterPartOperator
)

from . UE4WS_ObjectSocket import (
    Props as objectSocketProps,
    PANEL as objectSocketPanel,
    Ops as objectSocketOperator
)

from . UE4WS_ObjectLOD import (
    Groups as objectLODPGroup,
    Props as objectLODProps,
    PANEL as objectLODPanel,
    Ops as objectLODOperator
)

from . UE4WS_ObjectControlRig import (
    PANEL as objectControlRigPanel,
    Ops as objectControlRigOperator
)

from . UE4WS_ObjectRetargetAnimation import (
    RETARGET_UL_BoneParentList,
    RETARGET_UL_BoneMapList,
    Groups as objectRetargetAnimationGroup,
    Props as objectRetargetAnimationProps,
    PANEL as objectRetargetAnimationPanel,
    Ops as objectRetargetAnimationOperator
)

from . UE4WS_ObjectSkeletonPreset import (
    PANEL as objectSkeletonPresetPanel,
    Ops as objectSkeletonPresetOperator
)

from . UE4WS_StaticMesh import (
    PANEL as staticMeshPanel,
    Ops as staticMeshOperator
)

from . UE4WS_StaticMeshFBXOption import (
    PANEL as staticMeshFBXOptionPanel
)

from . UE4WS_StaticMeshUnrealEngine import (
    PANEL as staticMeshUnrealEnginePanel
)

from . UE4WS_Character import (
    PANEL as characterPanel,
    Ops as characterOperator
)

from . UE4WS_CharacterFBXOption import (
    PANEL as CharacterFBXOptionPanel
)

from . UE4WS_CharacterUnrealEngine import (
    PANEL as CharacterUnrealEnginePanel
)

from . UE4WS_Animation import (
    ANIM_UL_actionList,
    Props as animationProps,
    PANEL as animationPanel,
    Ops as animationOperator
)

from . UE4WS_AnimationFBXOption import (
    PANEL as AnimationFBXOptionPanel
)

from . UE4WS_AnimationUnrealEngine import (
    PANEL as AnimationUnrealEnginePanel
)

from . UE4WS_Credit import (
    PANEL as creditPanel
)

from . remote_execute import (
    RemoteExecution
)

from . misc import (
    PopUpWindow
)

remote_exec = RemoteExecution()

AR_UE4WS_PropsArray = []
# extend property to array, make sure you add from here
for x in [
    exportOptionImportAssetsProps,
    objectCustomCollisionProps,
    objectCharacterPartProps,
    objectSocketProps,
    objectLODProps,
    objectRetargetAnimationProps,
    animationProps
    ]:
    AR_UE4WS_PropsArray.extend(x)

AR_UE4WS_OperatorArray = []
# extend operator to array, make sure you add from here
for x in [
    PreferencesGroup,
    PreferencesOperator,
    objectOperator,
    objectCustomCollisionOperator,
    objectCharacterPartOperator,
    objectSocketOperator,
    objectLODPGroup,
    objectLODOperator,
    objectControlRigOperator,
    objectRetargetAnimationGroup,
    objectRetargetAnimationOperator,
    objectSkeletonPresetOperator,
    exportOptionOperator,
    exportOptionImportAssetsGroup,
    exportOptionImportAssetsOperator,
    staticMeshOperator,
    characterOperator,
    animationOperator
    ]:
    AR_UE4WS_OperatorArray.extend(x)

AR_UE4WS_classes = (
    # addon preferences
    Preferences,
    # Export Option
    exportOptionPanel,
    IMPORTASSETS_UL_AssetsList,
    exportOptionImportAssetsPanel,
    exportOptionImportAssetsFBXOptionPanel,
    exportOptionImportAssetsUnrealEnginePanel,
    # Object
    objectPanel,
    objectCustomCollisionPanel,
    objectCharacterPartPanel,
    objectSocketPanel,
    objectLODPanel,
    objectSkeletonPresetPanel,
    objectControlRigPanel,
    RETARGET_UL_BoneParentList,
    RETARGET_UL_BoneMapList,
    objectRetargetAnimationPanel,
    ## Static Mesh
    staticMeshPanel,
    staticMeshFBXOptionPanel,
    staticMeshUnrealEnginePanel,
    ## Character
    characterPanel,
    CharacterFBXOptionPanel,
    CharacterUnrealEnginePanel,
    ## Animation
    ANIM_UL_actionList,
    animationPanel,
    AnimationFBXOptionPanel,
    AnimationUnrealEnginePanel,
    ## Credit
    creditPanel,
    ## Misc.
    PopUpWindow
)

TypeProps = {
    "scene": bpy.types.Scene,
    "object": bpy.types.Object,
    "action": bpy.types.Action,
    "armature": bpy.types.Armature,
    "mesh": bpy.types.Mesh
}

@persistent
def resetVariable(scene):
    """Reset some variable and disconnect from unreal engine after load blend file"""
    preferences = bpy.context.preferences.addons[__package__].preferences
    preferences.skeleton.clear()
    preferences.CHAR_CharacterSkeleton = "NEW"
    preferences.ANIM_CharacterSkeleton = "NONE"
    remote_exec.stop()

    for P in AR_UE4WS_PropsArray:
        if hasattr(TypeProps.get(P.get("type")), P.get("name")) and P.get("resetVariable", False):
            typeProp = P.get("type")
            if (typeProp == "scene"):
                bpy.context.scene.property_unset(P.get("name"))
            elif (typeProp == "object"):
                for obj in bpy.data.objects:
                    obj.property_unset(P.get("name"))
            elif (typeProp == "action"):
                for action in bpy.data.actions:
                    action.property_unset(P.get("name"))
            elif (typeProp == "armature"):
                for armature in bpy.data.armatures:
                    armature.property_unset(P.get("name"))
            elif (typeProp == "mesh"):
                for armature in bpy.data.meshes:
                    armature.property_unset(P.get("name"))

def register():

    for X in AR_UE4WS_OperatorArray:
        if hasattr(X, "remote"):
            X.remote = remote_exec
        if hasattr(X, "addonVersion"):
            X.addonVersion = bl_info["version"]
        register_class(X)

    for P in AR_UE4WS_PropsArray:
        setattr(TypeProps.get(P.get("type")), P.get("name"), P.get("value"))

    for C in AR_UE4WS_classes:
        if hasattr(C, "remote"):
            C.remote = remote_exec
        if hasattr(C, "addonVersion"):
            C.addonVersion = bl_info["version"]
        register_class(C)

    bpy.app.handlers.load_post.append(resetVariable)

def unregister():

    for X in reversed(AR_UE4WS_OperatorArray):
        unregister_class(X)

    for P in reversed(AR_UE4WS_PropsArray):
        delattr(TypeProps.get(P.get("type")), P.get("name"))

    for C in reversed(AR_UE4WS_classes):
        unregister_class(C)

    bpy.app.handlers.load_post.remove(resetVariable)